### ISSUE TITLE
Merge Slave database back into master

### ISSUE THREAD

**Issue by IanMayo** [debrief/pepys-import#375]
Analysts may deploy to an offline workplace, and spend a period collecting data into a local instance of Pepys.

On their return, they need to be able to merge the new data back into the master database.

The new data may be in the `Reference`, `Metadata` or `Measurements` areas of the database (though it's most likely in the `Measurements` area - that's the point).

Note: for PostGIS we're already using GUIDs for row ids.  But, for SQLite I believe we're just using a counter. The counter is very convenient, but it won't work in this `master/slave` database management challenge.

**Comment by robintw** [comment]
One key question about this (just noting it here while I think of it), is whether it is expected that the master database will have additions to it during the period that the slave database is being used? I assume it is, and if so, then we would need to try and match up all of the metadata that was added on the two separate databases.

For example, the slave database could add `PLATFORM-X` and it would get a random GUID, and the master database could also add `PLATFORM-X` and it would get a different GUID. When merging we'd need to combine these into one entry (presumably the entry already in the master db) and alter the foreign keys on things like Sensors and States to point to the correct entry. This would also have to assume that a Platform entry is unique when considering name, nationality and platform type together (but that's something I'm going to make another issue about sometime, as I will need to assume that for some of the other changes I'm doing).

**Comment by IanMayo** [comment]
> master/slave conflicts
Yes, we will have to handle that _eventually_. 

In this instance, it won't be a great worry, since it's unlikely that the `master` will get any updates.

**Comment by IanMayo** [comment]
Aah, there _may_ be another troublesome issue.

They _may_ want to have periodic updates of the `slave` back to the `master`.  Aah, that's ok.  We only copy the data over if it's not already present.

**Comment by robintw** [comment]
Ah - having just seen your answer, if we can guarantee that master has no updates during the period the slave is away, then it will be fairly easy to merge them.

However, the implementation of the merge would have to change almost entirely if we then wanted to support updates on master while the slave is away - so it might be worth implementing it that way to start with - depending how much of an issue this is likely to be in future (I can imagine when Pepys is in operational use back on land, then there will be lots of updates while someone is away at sea).

My other concern is that if there are *any* updates to the master db while the slave is away, and we merged in a way that assumed there wasn't, then we'd end up with a lot of invalid data in the database: not just duplicated data, but actually wrong data. Obviously this isn't good!

**Comment by IanMayo** [comment]
Maybe:
* Phase one: reject merge if potential conflict identified
* Phase two: resolve conflicts

**Comment by robintw** [comment]
Sounds sensible.

I'll come up with some outline ideas for how to implement and then discuss with @BarisSari to work out a sensible way forward.

**Comment by robintw** [comment]
A few other questions while I plan this:

1. Could the master and slave dbs have had the same State/Contact data imported into them? Eg. could both the land and sea-based systems have received the same datafile and imported it? We can probably check that relatively easily as we keep track of which datafile each measurement comes from.

2. Further to the above, could they have imported the same data from a different datafile? Eg. a GPS sensor was recording for 2 days, and the master db imported a datafile containing the whole 2-day session, whereas the slave imported a datafile containing just the first day's data - so the actual numbers would be exactly the same, but the datafile name, hash and size would be different.

**Comment by IanMayo** [comment]
1.  Yes, it's possible
2. I think we can accept this duplication.  It's almost quite likely:  there may be one GPS, but multiple systems storing that location data & data gets stored from those multiple systems.

**Comment by robintw** [comment]
I've got a preliminary version of database merging working currently for the reference and metadata tables (and should be working soon for the measurements tables too). Along the way I've come up with a few questions, primarily around how we identify different or matching rows in the two databases.

For example, consider this scenario: the master database is exported to a slave database, and then the slave database is taken away to sea. During the time the databases are separated, each database has a platform called 'HMS Dolphin' added to it. This will end up with a row in the Platforms table in each database - and when we merge the databases, we want to ensure that this ends up as one single row. This raises various questions. A lot of this is around establishing uniqueness conditions (#379) - but a number of these questions go further than that. Specifically:

1. At the moment, the implementation assumes that all fields in the table must match for two separate records to be considered identical. For a Platform, this means that not only the platform name, type and nationality must match, but also the pennant number, trigraph and so on. Is this ok for a first implementation? We can alter this later to restrict to certain fields.

2. Should we consider privacy in the fields that have to be matched? For example, if the master has `HMS Dolphin` with a privacy of `Secret` and the slave has `HMS Dolphin` with a privacy of `Open`, what should we do? Do we assume they're the same platform? Do we merge and update with the most restrictive sort of privacy? If so, then we'll have to add extra columns to the Privacies table to define what order the privacies come in - and make that work across the codebase.

3. Should we consider empty fields when matching? For example if the master has `HMS Dolphin` with no data in the pennant number or trigraph fields, but the slave has `HMS Dolphin` with data in those fields - are they considered a match? And if so, do we then update the master with the extra data from the slave?

4. I hate to say this, as it will make things a lot more complicated, but do we need to go down the line of a sort of command-line resolver for the merge? For example, if the master database has `HMS Dolphin` and the slave has `Dolphin`, then we might want the user to be able to say these should be merged? I can see this getting quite complicated though - and the user may end up having to answer lots of questions. An alternative to this is that we just provide functionality in Pepys Admin to merge two objects together (ie. two platforms, two sensors, two nationalities) as a way of correcting mistakes that may have been made during the merge process (or during the import process itself - it sounds like quite a useful feature to have).

The same sort of questions that I've listed above will apply for other metadata tables too.

**Comment by IanMayo** [comment]
1. Hopefully the `platform uniqueness` answer will resolve this.  I'm expecting it to be `Name`, `Nationality`, `Pennant Number`. We will only need to check these.  (I've just chased a confirmation on the above)

2. Once we've established the uniqueness test, we merge, and use the highest privacy.  My _current_ intention is that they're added to the system in ascending order.  Aah, but we don't have integer IDs to use for comparison.  Yes, we should add an `Order` column, that we also use from the `import CSV reference data`. 

3. Yes, once we have established a match, we should merge the data.  We've identified how to handle conflicting privacy values.  Where value is missing, we can just merge them.  Where values clash, _sadly_ I think we have to reject merge.  Hmm, maybe not.  Actually, let's stick with the `Master` value. 

4. Yes, I'm quite happy with the latter - providing tools to (tidily) merge two objects.  This would be assisted if we produced a `Merge_report` that listed new metadata (or reference data) added to the database.  Any potential duplicates _Should_ leap out at the analyst.
